### PR TITLE
implement InterpretHealth for resource interpreter

### PR DESCRIPTION
**Pull request by zhuwint**
/kind feature

**What this PR does / why we need it**:

Implement InterpretHealthy method for resource interpreter

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```
`resource interpreter framework`: now able to interpret the resource health state.
```

### CODE REVIEW

File: pkg/resourceinterpreter/interpreter.go [hunk 1]
```diff
@@ -198,3 +201,22 @@ func (i *customResourceInterpreterImpl) ReflectStatus(object *unstructured.Unstr
...
+	if err != nil {
+		return
+	}
+	if hookEnabled {
+		return
+	}
+
+	healthy, err = i.defaultInterpreter.InterpretHealthy(object)
+	return
+}
```
Comment (duanmengkk) [inline pkg/resourceinterpreter/interpreter.go:222]: Add a new line of the end

Comment (zhuwint) [reply pkg/resourceinterpreter/interpreter.go]: OK

File: pkg/resourceinterpreter/interpreter.go [hunk 2]
```diff
@@ -42,6 +42,9 @@ type ResourceInterpreter interface {
 	// ReflectStatus returns the status of the object.
 	ReflectStatus(object *unstructured.Unstructured) (status *runtime.RawExtension, err error)
 
+	// InterpretHealthy returns the healthy check results of the object.
```
Comment (XiShanYongYe-Chang) [inline pkg/resourceinterpreter/interpreter.go:45]: // InterpretHealthy returns the healthy check result of the object.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Looks good~

We also need to update the doc file: `docs/userguide/customizing-resource-interpreter.md`.

**Comment by zhuwint** [comment]
The default healthy check of deployment refers to:
https://github.com/kubernetes/kubernetes/blob/bd9444c1cfd438690dc21c31404dd79ab60dc6cb/pkg/controller/deployment/rolling.go#L171-L174

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Good job!
/lgtm

File: pkg/resourceinterpreter/defaultinterpreter/default.go [hunk 3]
```diff
@@ -123,3 +126,14 @@ func (e *DefaultInterpreter) ReflectStatus(object *unstructured.Unstructured) (s
...
 }
+
+// InterpretHealthy returns the healthy check result of the object.
+func (e *DefaultInterpreter) InterpretHealthy(object *unstructured.Unstructured) (bool, error) {
+	handler, exist := e.healthyHandlers[object.GroupVersionKind()]
+	if exist {
+		return handler(object)
+	}
+
+	// for resource types that don't need to check healthy, return true directly
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/defaultinterpreter/default.go:137]: We can't assume the resource's default to `healthy`.
For resource types that we don't know, we should return error.

Comment (zhuwint) [reply pkg/resourceinterpreter/defaultinterpreter/default.go]: Thanks for pointing out, it has been modified in my latest commit.

File: pkg/resourceinterpreter/defaultinterpreter/healthy.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+func interpretDeploymentHealth(object *unstructured.Unstructured) (bool, error) {
+	deploy := &appsv1.Deployment{}
+	err := helper.ConvertToTypedObject(object, deploy)
+	if err != nil {
+		return false, err
+	}
+
+	healthy := deploy.Status.AvailableReplicas == *deploy.Spec.Replicas
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/defaultinterpreter/healthy.go:28]: I'd prefer the rule defined at [ArgoCD](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#deployment-replicaset-statefulset-daemonset):
```
Observed generation is equal to desired generation.
Number of updated replicas equals the number of desired replicas.
```

**Review by RainbowMango — COMMENTED** [review]
Looks great! Thanks @zhuwint 
Just one nit, otherwise lgtm.

File: pkg/resourceinterpreter/defaultinterpreter/healthy.go (hunk 4 above)
Comment (RainbowMango) [reply pkg/resourceinterpreter/defaultinterpreter/healthy.go:28]: Please add a unit test for this.

File: pkg/resourceinterpreter/defaultinterpreter/healthy.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+func interpretDeploymentHealth(object *unstructured.Unstructured) (bool, error) {
+	deploy := &appsv1.Deployment{}
+	err := helper.ConvertToTypedObject(object, deploy)
+	if err != nil {
+		return false, err
+	}
+
+	healthy := deploy.Status.UpdatedReplicas == *deploy.Spec.Replicas
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/defaultinterpreter/healthy.go:28]: https://github.com/karmada-io/karmada/pull/2329#discussion_r941324619

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**